doc/making-a-visualizer.md

```diff
@@ -20,187 +214,78 @@ the following data and methods:
-
-## Exporting a visualizer
-
-The engine expects visualizers to be packaged in `visualizerExportModules`
-which take as their arguments the name of the visualizer which is displayed in
-the UI. You can also reference the name of the visualizer to maintain
-consistency, though the `Differences` sets it explicitly.
-
-The other arguments in the export module are the visualizer itself
-(`Differences` in the example case, which is the name of the class we
-created), and a description.
-
-If you place the file containing your visualizer class definition with the
-export module in the folder name `Visualizers`, the engine will automatically
-package it up and include it in the list of available visualizers.
-
-There is no compiling needed. Simply place your file in the appropriate folder
-and run the app with `npm run dev`. The JavaScript will be compiled at
+   details, and you can just implement the `checkParameters()` method, which
+   only has to return a `ValidationStatus` that indicates whether the
+   parameter values are sensible.
+6. `show()`: Start or resume display of the visualization.
+7. `stop()`: Pause display of the visualization. Don't erase any visualization
+   produced so far or otherwise clean up the visualizer.
+8. `depart()`: Throw out the visualization, release its resources, remove its
+   injected DOM elements, and do any other required cleanup. After this call,
+   the visualizer must support `inhabit()` being called again, perhaps with a
+   different div, to re-initialize the visualization.
+
+### Where to put your visualizer
+
+When a visualizer is ready for Numberscope users, place the file containing
+its class definition and export module in the folder `src/visualizers`. When
+the Frontscope client runs, it'll find your visualizer and compile it at
 runtime.
 
-## Handling errors
+The visualizers in `src/visualizers` can be tested with the usual
+`npm run dev` call.
+
+As discussed [earlier](#test-your-visualizer-on-the-workbench), visualizers
+that aren't ready for Numberscope users should go in
+`src/visualizers-workbench`. You can load them for testing by calling
+`npm run dev:workbench`.
+
+### How to document your visualizer
+
+The p5 Template visualizer and the Differences visualizer both follow our
+documentation conventions.
+
+#### Name, thumbnail, and description
+
+In a documentation block just before the visualizer class definition, put a
+level-1 heading with the visualizer's name, followed by a thumbnail image and
+a description.
+
+To make the thumbnail, take a screenshot of your visualizer and put it in the
+`src/assets/img<VisualzierName>` directory. Set the `width` property in the
+`img` tag so that the image displays with a height of 320&nbsp;px or less.
+(You'll have to calculate the necessary width; setting the `height` property
+won't work.) Put the image tag inside a link to the image file.
+
+Below the thumbnail, describe what the visualizer does.
+
+### Parameters
```

Comment:
Oops, this has gone back to a level-3 heading itself (`###`) but it needs to definitely be seen as a child of "How to document your visualizer". So it and its sibling "Controls" need at least to be level-4 headings. But in fact I worry a bit that in the theme we are using, level 4 is not visually distinctive enough from level 3, and for expediency of making sure this sectioning is clear we should just make all of the siblings here level 5.  While we are on the topic of headings, it is not clear to me that the "Get started" heading/section is doing much. Maybe just delete it and promote its children up a level? And finally, I would suggest moving the "Document" section before "Test" -- if you aren't documenting your visualizer until you are testing it, you are documenting too late. Generally we should recommend documenting _before_ implementing. I definitely find it leads to better code. So indeed, you might want to insert a super brief "Implement" after "Document" and before "Test" -- it could just say "See the p5 visualizer section below for details on what properties and methods you need and what they are for."

Sorry this comment now has suggestions sort of throughout this file, but as they all occurred to me as I was considering these headings I thought it was best to get them down rather than take the time to write them all down and then have a bunch of conversations on these things scattered around this file. Thanks for understanding.

Reply:
> it is not clear to me that the "Get started" heading/section is doing much.

I intended "Get started" as a quick-start overview of the whole document. Its level-4 subsections reflect the level-3 subsections of the rest of the document:

Summary | Full
---|--
Choose a graphics framework | p5 visualizers (and other base class docs)
Test your visualizer | Where to put your visualizer
Document your visualizer | How to document your visualizer

Maybe the names could match up better?

> I would suggest moving the "Document" section before "Test" -- if you aren't documenting your visualizer until you are testing it, you are documenting too late.

My goal was to hand the reader whatever information they're reaching for at the moment. I figure that if you want to make a visualizer, your questions will probably go in this order:

Question | Answer
---|---
What do I type into my text editor? | Follow the "p5 visualizers" section
How do I see what I've done? | Run it on the workbench
How do I explain what I've done? | Here are our documentation conventions

If the wording of the headings didn't suggest this to you, we should revise them.

> Generally we should recommend documenting before implementing.

Could you clarify what that means to you? It's hard for me to imagine writing polished documentation for a feature when all I have is a paper sketch, with no idea whether I'll be able to turn it into working code.